ringio/io.go

```diff
@@ -69,7 +71,7 @@ func (b *Bounded) Read(p []byte) (n int, err error) {
 	defer b.Unlock()
 
 	if b.r.Empty() {
-		return 0, nil
+		return 0, io.EOF
```

Comment:
Oooh, yep, verrrry good catch. However, the ring buffer being empty at the point when the read happens doesn't mean it's terminally empty. From [the docs to io.Reader](https://pkg.go.dev/io@go1.19.1#Reader):

> Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0. Callers should treat a return of 0 and nil as indicating that nothing happened; in particular it does not indicate EOF.

Discouraged, yep, but I believe these are the semantics we're dealing with: Nothing happened but the ring might be readable in the future.

Reply:
To that end,
```suggestion
		return 0, nil
```

Reply:
I think it would be reasonable to add a Close method to Ring, so that any future reads can return `0, io.EOF` though!

Reply:
@antifuchs I changed this because your existing implementation results in a broken `io.Reader` implementation, which is to say it will work very poorly with any implementation that reads in a loop, and uses an error as the terminating case. Specifically, it will result in a tight loop doing nothing.

Examples:

- [io.ReadAll](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/io/io.go;l=654)
- [io.Copy](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/io/io.go;l=405) (and equivalents)
- [bytes.Buffer.ReadFrom](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/bytes/buffer.go;l=197) (also used by `io.Copy` and similar)

Presumably, this is one reason why it is discouraged. The `io.EOF` error is simply a sentinel value, indicating that there is no more data to be read. Your current implementation is simply a fixed buffer, and logically is never opened or closed, much the same way as `bytes.Buffer` is never opened or closed.

Supporting blocking behavior would be best implemented by the caller, IMO, e.g. using [io.Pipe](https://pkg.go.dev/io#Pipe), or [net.Pipe](https://pkg.go.dev/net#Pipe). Both those implementations use the same style implementation (using a ping-pong pattern, via channels), in order to mitigate deadlocks, which IIRC persisted in `io.Pipe` for a very long time. Implementing that functionality _within_ the ring buffer (instead of just letting the caller implement it, fairly trivially) would significantly impact it's fit for (what I would consider) this implementation's core use cases. My own use case, included.

Reply:
As an example of an implementation that doesn't follow a strict "returns io.EOF on close" pattern, consider [tar.Reader.Read](https://pkg.go.dev/archive/tar#Reader.Read), which returns `io.EOF` to bound each _archived_ file, of which they may be more than one, within the input stream.

Reply:
Hm, yeah you make a good point there. I feel that the go stdlib could use some clarification about what the "you're at the end" behavior of an io.Read should be - but if we can just re-use an existing Read object (which it seems to pointedly avoid discussing), the behavior you introduced makes sense - and if it fixed io.ReadAll, even better!